toolchain/lower/file_context.cpp

```diff
@@ -146,6 +160,240 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+    SemIR::SpecificId specific_id1, SemIR::SpecificId specific_id2,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>> set_of_pairs) -> bool {
+  if (specific_id1.index > specific_id2.index) {
+    std::swap(specific_id1.index, specific_id2.index);
+  }
+  std::pair<SemIR::SpecificId, SemIR::SpecificId> pair = {specific_id1,
+                                                          specific_id2};
+  return set_of_pairs.Contains(pair);
+}
+
+auto FileContext::CoalesceEquivalentSpecifics() -> void {
+  for (auto& specifics : lowered_specifics_) {
+    // i cannot be unsigned due to the comparison with a negative number when
+    // the specifics vector is empty.
+    for (int i = 0; i < static_cast<int>(specifics.size()) - 1; ++i) {
+      // This specific was already replaced, skip it.
+      if (equivalent_specifics_[specifics[i].index].has_value() &&
+          equivalent_specifics_[specifics[i].index] != specifics[i]) {
+        specifics[i] = specifics[specifics.size() - 1];
+        specifics.pop_back();
+        --i;
+        continue;
+      }
+      // TODO: improve quadratic behavior by using a single hash based on
+      // lowered_specifics_type_fingerprint_ and function_common_fingerprint.
+      for (int j = i + 1; j < static_cast<int>(specifics.size()); ++j) {
+        // This specific was already replaced, skip it.
+        if (equivalent_specifics_[specifics[j].index].has_value() &&
+            equivalent_specifics_[specifics[j].index] != specifics[j]) {
+          specifics[j] = specifics[specifics.size() - 1];
+          specifics.pop_back();
+          --j;
+          continue;
+        }
+
+        // The two specifics are not equivalent due to the function type info,
+        // stored in lowered_specifics_types.
+        // Mark non-equivalance in case it can be reused to short-cut another
+        // path and continue the search for other equivalences.
+        if (!AreFunctionTypesEquivalent(specifics[i], specifics[j])) {
+          (void)PairInserted(specifics[i], specifics[j],
+                             non_equivalent_specifics_);
+          continue;
+        }
+
+        Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>
+            visited_equivalent_specifics;
+        (void)PairInserted(specifics[i], specifics[j],
+                           visited_equivalent_specifics);
+        // Function type information matches, check usages inside function body
+        // that are dependent on the specific. This information has been stored
+        // in lowered_states while lowering each function body.
+        if (AreFunctionBodiesEquivalent(specifics[i], specifics[j],
+                                        visited_equivalent_specifics)) {
+          // This is a Set, because it's possible, when processing equivalences
+          // found, that the same function is found to need replacement multiple
+          // times, when a new ("better") canonical is found after an initial
+          // replacement.
```

Comment:
This is no longer a set, right? Maybe:
```suggestion
          // When processing equivalences, we may change the canonical specific
          // multiple times, so we don't delete and update the canonical until the end.
          // TODO: Try to address this without repeated replacements. 
```

Note though, this is also where I'm not confident that this is the right way to address this issue. So the TODO.

Reply:
Updating comment to reflect the "not deleting, but replacing."